File: whatsrisky/runners/gitleaks.py
"""gitleaks - hardcoded secrets in the working tree and in git history."""

from __future__ import annotations

import json
import re

from ..models import Finding, Severity
from ..util import pattern_to_regex, relative, run, run_streaming, truncate
from .base import Runner

TRIPLE = "'" * 3
# gitleaks logs as "11:07PM INF message"; only the message is worth showing.
_LOG_PREFIX = re.compile(r"^\d{1,2}:\d{2}(AM|PM)?\s+(INF|WRN|ERR|DBG|TRC)\s+", re.IGNORECASE)

# Rules that match on shape/entropy alone produce more false positives than
# provider-specific rules, so they land one notch lower.
_GENERIC_RULES = {
    "generic-api-key",
    "generic-api-token",
    "high-entropy-string",
    "jwt",
    "private-key",
}
_NON_PROD_HINT = re.compile(
    r"(^|/)(tests?|spec|specs|fixtures?|examples?|samples?|mocks?|__tests__|testdata)(/|$)"
    r"|\.(example|sample|template|dist|md|mdx|rst|txt)$",
    re.IGNORECASE,
)


class GitleaksRunner(Runner):
    name = "gitleaks"
    binary = "gitleaks"
    category = "Secret"
    install_hints = {
        "darwin": "brew install gitleaks",
        "linux": "download a release from https://github.com/gitleaks/gitleaks/releases",
        "windows": "scoop install gitleaks",
        "default": "https://github.com/gitleaks/gitleaks/releases",
    }

    def version(self) -> str:
        res = run([self.binary, "version"], timeout=30)
        raw = (res.stdout or res.stderr).strip().splitlines()
        return f"gitleaks {raw[0].strip()}" if raw else ""

    def _version_tuple(self) -> tuple[int, ...]:
        match = re.search(r"(\d+)\.(\d+)\.(\d+)", self.version())
        return tuple(int(g) for g in match.groups()) if match else (0, 0, 0)

    def _modes(self) -> list[str]:
        """Which gitleaks passes to run: working tree, git history, or both."""
        from ..util import is_git_repo

        if self.config.diff_range:
            return ["git"]  # only the commits in the range matter
        mode = self.config.gitleaks_mode
        if mode == "dir":
            return ["dir"]
        if mode == "git":
            return ["git"]
        return ["dir", "git"] if is_git_repo(self.config.target) else ["dir"]

    def _exclude_config(self) -> str:
        """gitleaks has no --exclude flag: paths are excluded through a config allowlist."""
        patterns = [p for p in (pattern_to_regex(x) for x in self.config.exclude) if p]
        if not patterns:
            return ""
        path = self.config.work_dir / "gitleaks-excludes.toml"
        lines = [
            "[extend]",
            "useDefault = true",
            "",
            "[[allowlists]]",
            'description = "whatsrisky excludes"',
            "paths = [",
        ]
        lines += ["  " + TRIPLE + p + TRIPLE + "," for p in patterns]
        lines.append("]")
        path.write_text("\n".join(lines) + "\n", encoding="utf-8")
        return str(path)

    def _argv(self, mode: str, report: str) -> list[str]:
        cfg = self.config
        modern = self._version_tuple() >= (8, 19, 0)
        common = [
            "--report-format",
            "json",
            "--report-path",
            report,
            "--exit-code",
            "0",
            "--no-banner",
            "--redact",
        ]
        scoped = ["--log-opts", cfg.diff_range] if (mode == "git" and cfg.diff_range) else []
        config_path = self._exclude_config()
        if config_path:
            scoped += ["--config", config_path]
        if modern:
            return [self.binary, mode, ".", *common, *scoped]
        argv = [self.binary, "detect", "--source", ".", *common, *scoped]
        if mode == "dir":
            argv.append("--no-git")
        return argv

    def scan(self):
        cfg = self.config
        findings: list[Finding] = []
        commands: list[str] = []
        stderr_all: list[str] = []
        seen: set[str] = set()

        for mode in self._modes():
            report = cfg.work_dir / f"gitleaks-{mode}.json"
            if report.exists():
                report.unlink()
            self.progress(f"scanning {'git history' if mode == 'git' else 'working tree'}")
            res = run_streaming(
                self._argv(mode, str(report)),
                cwd=cfg.target,
                timeout=cfg.gitleaks_timeout,
                on_stderr=lambda line: self.progress(_LOG_PREFIX.sub("", line.strip())),
            )
            commands.append(res.command)
            stderr_all.append(res.stderr)
            if not report.exists():
                if res.returncode not in (0, 1):
                    stderr_all.append(f"[whatsrisky] gitleaks {mode} exit {res.returncode}")
                continue
            text = report.read_text(encoding="utf-8", errors="replace").strip()
            if not text:
                continue
            try:
                entries = json.loads(text)
            except ValueError as exc:
                stderr_all.append(f"[whatsrisky] unparsable gitleaks {mode} report: {exc}")
                continue
            for item in entries or []:
                finding = self._to_finding(item, mode)
                key = finding.fingerprint
                if key in seen:
                    continue
                seen.add(key)
                findings.append(finding)

        if not commands:
            raise RuntimeError("gitleaks did not run")
        return findings, " && ".join(commands), "\n".join(stderr_all)

    def _to_finding(self, item: dict, mode: str) -> Finding:
        rule = item.get("RuleID", "")
        rel = relative(item.get("File", ""), self.config.target)
        severity = Severity.CRITICAL
        if rule in _GENERIC_RULES:
            severity = Severity.HIGH
        if rel and _NON_PROD_HINT.search(rel):
            severity = Severity.HIGH if severity is Severity.CRITICAL else Severity.MEDIUM

        commit = item.get("Commit", "")
        where = "git history" if commit else "working tree"
        detail = [
            f"gitleaks rule `{rule}` matched a likely credential in the {where}.",
            f"Description: {item.get('Description', '')}",
        ]
        if commit:
            detail.append(
                f"Commit {commit[:12]} by {item.get('Author', '?')} <{item.get('Email', '?')}> "
                f"on {item.get('Date', '?')}"
            )
        if item.get("Entropy"):
            detail.append(f"Entropy: {item['Entropy']}")

        return Finding(
            tool=self.name,
            severity=severity,
            title=truncate(f"Hardcoded secret: {rule or 'unknown rule'}", 140),
            description="\n".join(d for d in detail if d.strip()),
            category="Secret" + ("/git-history" if commit else "/working-tree"),
            rule_id=rule,
            file=rel,
            line=item.get("StartLine") or None,
            end_line=item.get("EndLine") or None,
            cwe=["CWE-798"],
            remediation=(
                "1) Treat the credential as compromised and rotate it at the provider. "
                "2) Remove it from the source and load it from a secret manager/env var. "
                "3) If it is in git history, purge it (git filter-repo / BFG) and force-push. "
                "4) Add a pre-commit gitleaks hook to prevent recurrence."
            ),
            snippet=truncate(item.get("Match", ""), 240),
            pass_name=mode,
            raw={"mode": mode, "commit": commit, "fingerprint": item.get("Fingerprint", "")},
        )

File: internal/runner/gitleaks.go
package runner

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"strings"

	"github.com/smagew/whatsrisky/internal/exclude"
	"github.com/smagew/whatsrisky/internal/model"
	"github.com/smagew/whatsrisky/internal/proc"
)

// Gitleaks - hardcoded secrets in the working tree and in git history.
type Gitleaks struct{ binaryRunner }

// NewGitleaks builds the runner.
func NewGitleaks(config Config) *Gitleaks {
	return &Gitleaks{binaryRunner{
		binary: "gitleaks",
		hints: installHints{
			"darwin":  "brew install gitleaks",
			"linux":   "download a release from https://github.com/gitleaks/gitleaks/releases",
			"windows": "scoop install gitleaks",
			"default": "https://github.com/gitleaks/gitleaks/releases",
		},
		config: config,
	}}
}

func (g *Gitleaks) Name() string { return "gitleaks" }

func (g *Gitleaks) Version() string {
	if version := proc.Version(g.binary, "version"); version != "" {
		return "gitleaks " + version
	}
	return ""
}

var (
	versionNumbers = regexp.MustCompile(`(\d+)\.(\d+)\.(\d+)`)
	// gitleaks logs as "11:07PM INF message"; only the message is worth showing.
	logPrefix = regexp.MustCompile(`(?i)^\d{1,2}:\d{2}(AM|PM)?\s+(INF|WRN|ERR|DBG|TRC)\s+`)

	// Rules that match on shape or entropy alone produce more false positives than
	// provider-specific ones, so they land one notch lower.
	genericRules = map[string]bool{
		"generic-api-key": true, "generic-api-token": true,
		"high-entropy-string": true, "jwt": true, "private-key": true,
	}
	nonProdPath = regexp.MustCompile(
		`(?i)(^|/)(tests?|spec|specs|fixtures?|examples?|samples?|mocks?|__tests__|testdata)(/|$)` +
			`|\.(example|sample|template|dist|md|mdx|rst|txt)$`)
)

func (g *Gitleaks) versionAtLeast(major, minor, patch int) bool {
	match := versionNumbers.FindStringSubmatch(g.Version())
	if len(match) != 4 {
		return false
	}
	have := [3]int{atoi(match[1]), atoi(match[2]), atoi(match[3])}
	want := [3]int{major, minor, patch}
	for i := range have {
		if have[i] != want[i] {
			return have[i] > want[i]
		}
	}
	return true
}

func atoi(text string) int {
	value := 0
	for _, char := range text {
		if char < '0' || char > '9' {
			return value
		}
		value = value*10 + int(char-'0')
	}
	return value
}

// modes decides which passes to run: the working tree, git history, or both.
func (g *Gitleaks) modes() []string {
	if g.config.DiffRange != "" {
		return []string{"git"} // only the commits in the range matter
	}
	switch g.config.GitleaksMode {
	case "dir":
		return []string{"dir"}
	case "git":
		return []string{"git"}
	}
	if isGitRepo(g.config.Target) {
		return []string{"dir", "git"}
	}
	return []string{"dir"}
}

func isGitRepo(dir string) bool {
	result, err := proc.Run([]string{"git", "rev-parse", "--is-inside-work-tree"},
		proc.Options{Dir: dir, Timeout: gitProbeTimeout})
	return err == nil && result.ExitCode == 0 && strings.TrimSpace(result.Stdout) == "true"
}

// excludeConfig writes the allowlist gitleaks needs: it has no --exclude flag, so
// paths are excluded through a config.
func (g *Gitleaks) excludeConfig() (string, error) {
	var patterns []string
	for _, pattern := range g.config.Exclude {
		if regex := exclude.PatternToRegex(pattern); regex != "" {
			patterns = append(patterns, regex)
		}
	}
	if len(patterns) == 0 {
		return "", nil
	}
	const triple = "'''"
	lines := []string{
		"[extend]", "useDefault = true", "",
		"[[allowlists]]", `description = "whatsrisky excludes"`, "paths = [",
	}
	for _, pattern := range patterns {
		lines = append(lines, "  "+triple+pattern+triple+",")
	}
	lines = append(lines, "]")

	path := filepath.Join(g.config.WorkDir, "gitleaks-excludes.toml")
	return path, WriteFile(path, strings.Join(lines, "\n")+"\n")
}

func (g *Gitleaks) argv(mode, report, configPath string) []string {
	common := []string{
		"--report-format", "json", "--report-path", report,
		"--exit-code", "0", "--no-banner", "--redact",
	}
	var scoped []string
	if mode == "git" && g.config.DiffRange != "" {
		scoped = append(scoped, "--log-opts", g.config.DiffRange)
	}
	if configPath != "" {
		scoped = append(scoped, "--config", configPath)
	}
	if g.versionAtLeast(8, 19, 0) {
		return append(append([]string{g.binary, mode, "."}, common...), scoped...)
	}
	argv := append(append([]string{g.binary, "detect", "--source", "."}, common...), scoped...)
	if mode == "dir" {
		argv = append(argv, "--no-git")
	}
	return argv
}

type gitleaksFinding struct {
	Description string  `json:"Description"`
	StartLine   int     `json:"StartLine"`
	EndLine     int     `json:"EndLine"`
	Match       string  `json:"Match"`
	File        string  `json:"File"`
	Commit      string  `json:"Commit"`
	Entropy     float64 `json:"Entropy"`
	Author      string  `json:"Author"`
	Email       string  `json:"Email"`
	Date        string  `json:"Date"`
	RuleID      string  `json:"RuleID"`
	Fingerprint string  `json:"Fingerprint"`
}

func (g *Gitleaks) Scan(progress func(string)) (Outcome, error) {
	config := g.config
	configPath, err := g.excludeConfig()
	if err != nil {
		return Outcome{}, fmt.Errorf("writing the gitleaks allowlist: %w", err)
	}

	var findings []model.Finding
	var commands, stderrs []string
	seen := map[string]bool{}

	for _, mode := range g.modes() {
		report := filepath.Join(config.WorkDir, "gitleaks-"+mode+".json")
		_ = os.Remove(report)
		where := "working tree"
		if mode == "git" {
			where = "git history"
		}
		progress("scanning " + where)

		result, err := proc.Run(g.argv(mode, report, configPath), proc.Options{
			Dir:     config.Target,
			Timeout: config.GitleaksTimeout,
			OnStderr: func(line string) {
				progress(logPrefix.ReplaceAllString(strings.TrimSpace(line), ""))
			},
		})
		if err != nil {
			return Outcome{}, err
		}
		commands = append(commands, result.Command())
		stderrs = append(stderrs, result.Stderr)

		raw, readErr := os.ReadFile(report)
		if readErr != nil {
			if result.ExitCode != 0 && result.ExitCode != 1 {
				stderrs = append(stderrs, fmt.Sprintf("[whatsrisky] gitleaks %s exit %d", mode, result.ExitCode))
			}
			continue
		}
		if strings.TrimSpace(string(raw)) == "" {
			continue
		}
		var entries []gitleaksFinding
		if err := json.Unmarshal(raw, &entries); err != nil {
			stderrs = append(stderrs, fmt.Sprintf("[whatsrisky] unparsable gitleaks %s report: %v", mode, err))
			continue
		}
		for _, entry := range entries {
			finding := g.toFinding(entry, mode)
			if seen[finding.Fingerprint()] {
				continue
			}
			seen[finding.Fingerprint()] = true
			findings = append(findings, finding)
		}
	}
	if len(commands) == 0 {
		return Outcome{}, fmt.Errorf("gitleaks did not run")
	}
	return Outcome{
		Findings: findings,
		Command:  strings.Join(commands, " && "),
		Stderr:   strings.Join(stderrs, "\n"),
	}, nil
}

// toFinding maps a gitleaks hit onto the shared scale. gitleaks has no severity of
// its own, so this is where one is assigned - and where the reasoning lives.
func (g *Gitleaks) toFinding(entry gitleaksFinding, mode string) model.Finding {
	rel := Relative(g.config.Target, entry.File)
	severity := model.Critical
	if genericRules[entry.RuleID] {
		severity = model.High
	}
	if rel != "" && nonProdPath.MatchString(rel) {
		if severity == model.Critical {
			severity = model.High
		} else {
			severity = model.Medium
		}
	}

	where := "working tree"
	category := "Secret/working-tree"
	if entry.Commit != "" {
		where = "git history"
		category = "Secret/git-history"
	}
	detail := []string{
		fmt.Sprintf("gitleaks rule `%s` matched a likely credential in the %s.", entry.RuleID, where),
		"Description: " + entry.Description,
	}
	if entry.Commit != "" {
		commit := entry.Commit
		if len(commit) > 12 {
			commit = commit[:12]
		}
		detail = append(detail, fmt.Sprintf("Commit %s by %s <%s> on %s",
			commit, orText(entry.Author, "?"), orText(entry.Email, "?"), orText(entry.Date, "?")))
	}
	if entry.Entropy != 0 {
		detail = append(detail, fmt.Sprintf("Entropy: %g", entry.Entropy))
	}

	finding := model.Finding{
		Tool:            "gitleaks",
		Severity:        severity,
		Title:           model.Truncate("Hardcoded secret: "+orText(entry.RuleID, "unknown rule"), 140),
		Description:     strings.Join(detail, "\n"),
		ScannerCategory: category,
		RuleID:          entry.RuleID,
		File:            rel,
		Line:            entry.StartLine,
		EndLine:         entry.EndLine,
		CWE:             []string{"CWE-798"},
		Remediation: "1) Treat the credential as compromised and rotate it at the provider. " +
			"2) Remove it from the source and load it from a secret manager/env var. " +
			"3) If it is in git history, purge it (git filter-repo / BFG) and force-push. " +
			"4) Add a pre-commit gitleaks hook to prevent recurrence.",
		Snippet: model.Truncate(entry.Match, 240),
		Pass:    mode,
		Raw:     map[string]string{"mode": mode, "commit": entry.Commit, "fingerprint": entry.Fingerprint},
	}
	finding.Normalize()
	return finding
}
